File: Go/day04/part2.go
package main

import (
	"adventofcode/packages/utils"
	"strings"
)

/*
This is a working solution but I have no idea if it's idomatic Go code.
I expect to come back to this after getting more experience with Go.
*/

func part2() int {
	content := utils.GetPuzzleInput("day04")
	lines := make([][][]int, 0)
	c := 0
	for _, line := range content {
		temp := strings.Split(line, ",")
		p := make([][]int, 0)

		for _, v := range temp {
			arr := strings.Split(v, "-")
			nums := utils.ToInts(arr)
			p = append(p, nums)
		}

		lines = append(lines, p)
	}

	for _, line := range lines {
		a1, a2, b1, b2 := line[0][0], line[0][1], line[1][0], line[1][1]

		if a2 >= b1 && b2 >= a1 {
			c++
		}
	}

	return c
}

File: TypeScript/day04/part2.ts
import { getPuzzleInput } from "../utils";
const input = getPuzzleInput("day04");

export default function part1(input: string) {
  const pairs = input
    .split("\n")
    .map((p) =>
      p.split(",").map((p) => p.split("-").map((item) => Number(item)))
    );
  let partialCount = 0;

  for (let pair of pairs) {
    const [[a1, a2], [b1, b2]] = pair;

    if (a2 >= b1 && b2 >= a1) {
      partialCount++;
    }
  }
  return partialCount;
}
